Document about wallet backup and restoration (#22523)
bitcoin/bitcoin: doc/managing-wallet.md

# Managing the Wallet

## 1. Backing Up and Restoring The Wallet

### 1.1 Creating the Wallet

Since version 0.21, Bitcoin Core no longer has a default wallet.

> Wallets can be created with the `createwallet` RPC. More information about this command may be found by running `bitcoin-cli help createwallet`.

Reviewer: I think it would be useful to provide instructions for creating a wallet via the GUI.

Author: Done.

The following command, for example, creates a descriptor wallet:

`$ bitcoin-cli -named createwallet wallet_name="wallet-01" descriptors=true`

The `descriptors` parameter can be omitted if the intention is to create a legacy wallet.

For now, the default type is the legacy wallet, but that should change in the near future.

By default, wallets are created in the `wallets` folder of the data directory, which varies by operating system, as shown below. The user can change the default by using the `-datadir` or `-walletdir` initialization parameters.

| Operating System | Default wallet directory                                    |
| -----------------|:------------------------------------------------------------|
| Linux            | `/home/<user>/.bitcoin/wallets`                             |
| Windows          | `C:\Users\<user>\AppData\Roaming\Bitcoin\wallets`           |
| macOS            | `/Users/<user>/Library/Application Support/Bitcoin/wallets` |

### 1.2 Encrypting the Wallet

The `wallet.dat` file is not encrypted by default and is, therefore, vulnerable if an attacker gains access to the device where the wallet or the backups are stored.

> The wallet's private key may be encrypted with the following command:

Reviewer 2: I think this can be addressed in a follow up, but I believe we generate a new HD seed when the wallet goes from unencrypted to encrypted, and likewise (on the lines below) when the passphrase is changed.  That means the user should create a new backup immediately after encrypting or changing the passphrase, otherwise any bitcoins received to the new seed cannot be recovered from the previous backups.

Reviewer 3: Yes, I mentioned that [here](https://github.com/bitcoin/bitcoin/pull/22523#discussion_r677926273) (actually quoted you) :)
And the text was changed from "...private key must be encrypted..." to "...private key may be encrypted...." but it is probably a good idea to include more information on why this can be dangerous.

Author: Added information about tradeoff between encrypted wallet and risk of forgotten passphrases.
Also added information about the new HD seed that is created when the wallet is encrypted or the passphrase changes.
These are really important considerations.

`$ bitcoin-cli -rpcwallet="wallet-01" encryptwallet "passphrase"`

Once encrypted, the passphrase may be changed with the `walletpassphrasechange` command.

`$ bitcoin-cli -rpcwallet="wallet-01" walletpassphrasechange "oldpassphrase" "newpassphrase"`

The argument passed to `-rpcwallet` is the name of the wallet to be encrypted.

Only the wallet's private key is encrypted. All other wallet information, such as transactions, is still visible.

The wallet's private key can also be encrypted in the `createwallet` command via the `passphrase` argument:

`$ bitcoin-cli -named createwallet wallet_name="wallet-01" descriptors=true passphrase="passphrase"`

Note that if the passphrase is lost, all the coins in the wallet will also be lost forever.

### 1.3 Backing Up the Wallet

> Wallets can be copied to another destination. This backup file should be stored on offline devices only, such as on a USB drive or an external hard drive.

Reviewer: It is potentially dangerous to just copy the wallet file. I don't think we should encourage users to just copy wallet files. Rather they should use the built in backup functionality.

Reviewer 4: Yes. I was trying to get at that with https://github.com/bitcoin/bitcoin/pull/22523#discussion_r680481394.

Author: It has been changed to make it clearer that the built in backup should be used.

If the wallet and backup are lost for any reason, the bitcoins related to this wallet will become permanently inaccessible.

> The `backupwallet` RPC or the `Backup Wallet` GUI menu item must be used to ensure the file is in a safe state when the copy is made. The destination parameter must include the name of the file. Otherwise, the command will return an error message like "Error: Wallet backup failed!" for descriptor wallets. If it is a legacy wallet, it will be copied and a file will be created with the default file name `wallet.dat`.

Reviewer: It would be useful to also mention how backups can be performed via the GUI.

Author: Done.

The original Bitcoin Core wallet was a collection of unrelated private keys. If a non-HD wallet had received funds to an address and then was restored from a backup made before the address was generated, then any funds sent to that address would have been lost because there was no deterministic mechanism to derive the address again.

Bitcoin Core [version 0.13](https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.13.0.md) introduced HD wallets with deterministic key derivation. With HD wallets, users no longer lose funds when restoring old backups because all addresses are derived from the HD wallet seed.

This means that a single backup is enough to recover the coins at any time. It is still recommended to make regular backups (once a week) or after a significant number of new transactions to maintain the metadata, such as labels. Metadata cannot be retrieved from a blockchain rescan, so if the backup is too old, the metadata will be lost forever.

Wallets created before version 0.13 are not HD and must be backed up every 100 keys used since the previous backup, or even more often to maintain the metadata.

### 1.5 Restoring the Wallet From a Backup

> To restore a wallet, an empty one must be created first. The backup file then rewrites the `wallet.dat` of this new wallet.

Reviewer: No, this is potentially dangerous. This entire section encourages a potentially dangerous restore process.

The correct process is to create a directory in the walletdir with a name that the user wants the wallet to be named, and then copy the wallet.dat file into that directory. There should not be any overwriting of anything.

Author: Creating a new folder is actually safer than overwriting it. Updated the section.

# Managing the Wallet

## 1. Backing Up and Restoring The Wallet

### 1.1 Creating the Wallet

Since version 0.21, Bitcoin Core no longer has a default wallet.

> Wallets can be created with the RPC command `createwallet`. More information about this command can be found running `bitcoin-cli help createwallet`.

Reviewer 4: "RPC command" is redundant (the "C" in RPC means "call")
```suggestion
Wallets can be created with the `createwallet` RPC. More information about this command may be found by running `bitcoin-cli help createwallet`.
```

The following command, for example, creates a descriptor wallet:

`$ bitcoin-cli -named createwallet wallet_name="wallet-01" descriptors=true`

The `descriptors` parameter can be omitted if the intention is to create a legacy wallet.

For now, the default type is the legacy wallet, but that should change in the near future.

> Wallets are created in the `wallets` folder of the data directory. The default folder varies by operating systems, as shown below. The user can also change the default by using the `-datadir` or `-walletdir` initialization parameters.

Reviewer 4: ```suggestion
By default, wallets are created in the `wallets` folder of the data directory, which varies by operating system, as shown below. The user can change the default by using the `-datadir` or `-walletdir` initialization parameters.
```

| Operating System | Default wallet directory                                    |
| -----------------|:------------------------------------------------------------|

> | Linux            | `home/<user>/.bitcoin/wallets`                              |

Reviewer 4: ```suggestion
| Linux            | `/home/<user>/.bitcoin/wallets`                             |
```

| Windows          | `C:\Users\<user>\AppData\Roaming\Bitcoin\wallets`           |
| macOS            | `/Users/<user>/Library/Application Support/Bitcoin/wallets` |

### 1.2 Encrypting the Wallet

The `wallet.dat` file is not encrypted by default and is, therefore, vulnerable if an attacker gains access to the device where the wallet or the backups are stored.

> The wallet may be encrypted with the following command:

Reviewer 4: ```suggestion
The wallet's private key may be encrypted with the following command:
```

`$ bitcoin-cli -rpcwallet="wallet-01" encryptwallet "passphrase"`

> The `encryptwallet` command is used only when the wallet has not been encrypted yet. Otherwise, the `walletpassphrasechange` command should be used.

Reviewer 4: ```suggestion
Once encrypted, the passphrase may be changed with the `walletpassphrasechange` command.
```

`$ bitcoin-cli -rpcwallet="wallet-01" walletpassphrasechange "oldpassphrase" "newpassphrase"`

The argument passed to `-rpcwallet` is the name of the wallet to be encrypted.

> The term "encrypt the wallet" used here is not very accurate. This command encrypts only the private key. All other wallet information, such as transactions, is still visible.

Reviewer 4: ```suggestion
Only the wallet's private key is encrypted. All other wallet information, such as transactions, is still visible.
```

> The wallet can also be encrypted in the `createwallet` command via the `passphrase` argument:

Reviewer 4: ```suggestion
The wallet's private key can also be encrypted in the `createwallet` command via the `passphrase` argument:
```

`$ bitcoin-cli -named createwallet wallet_name="wallet-01" descriptors=true passphrase="passphrase"`

Note that if the passphrase is lost, all the coins in the wallet will also be lost forever.

### 1.3 Backing Up the Wallet

> Wallets can be safely copied to another destination. This backup file should be stored on offline devices only, such as on a USB drive or an external hard drive.

Reviewer 5: Not sure about this. Offline device is better than Online device but still not safe.

Reviewer 3: If not offline or online, where?

Reviewer 6: I think he meant that not _any_ offline device is safe. But I would say the lack of security @prayank23 pointed out on some offline devices may be out of the scope of this documentation. 

What I understand of `Wallets can be safely copied` is it refers to the fact that the funds will be recoverable just backuing up that file. Then he is emphasizing the fact that online storage is a bad idea, which I think we all agree.

Reviewer 5: > If not offline or online, where?

Offline is okay but USB drives cannot be trusted and in some cases external hard drive as well. So if I had to rephrase the above thing and add few more options, it would be something like this:

_Wallets should be safely copied to another destination. This backup file can be stored on multiple offline devices that you trust will work in emergency, never connected to online devices and no malware. Test your backups regularly to avoid any issues in future. Also printing the output for `dumpwallet` on a sheet of paper could be helpful in some cases._

> I think he meant that not any offline device is safe. But I would say the lack of security @prayank23 pointed out on some offline devices may be out of the scope of this documentation. 

Agree. Can't go in to details but this repository is used by lot of users with different levels of technical expertise so nobody should trust one file on some USB.

Reviewer 3: Sure, there are security issues with USB devices (propriety firmware, etc) - but I think it is out of scope for this doc (for example, we don't tell our users to only use Linux/Power ISA devices).

I do like the idea of including `dumpwallet` as an option - especially with [this](https://github.com/bitcoin/bitcoin/pull/11803) PR hopefully being merged (although we should inform that this option would lose metadata)

Author: The document originally had two sections on `dumpwallet` and `importwallet`. But it was removed for the reason cited in https://github.com/bitcoin/bitcoin/pull/22523#discussion_r674488255.
From the reviews, it seems that there is a preference for a shorter and less verbose document.
But I agree that `dumpwallet` backup and security notes are a good follow-up.

Reviewer 3: Makes sense.  Maybe a `listdescriptors` follow up too https://github.com/bitcoin/bitcoin/pull/21500

Author: Added the suggested security considerations.

> Wallets can be safely copied to another destination. This backup file should be stored on offline devices only, such as on a USB drive or an external hard drive.

Reviewer 4: ```suggestion
Wallets can be copied to another destination. This backup file should be stored on offline devices only, such as on a USB drive or an external hard drive.
```

It may be clearer for this paragraph and the `backupwallet` one to be together so the user doesn't interpret "copy" differently.

If the wallet and backup are lost for any reason, the bitcoins related to this wallet will become permanently inaccessible.

> The `bitcoin-cli backupwallet "destination"` RPC command or `Backup Wallet` GUI menu item must be used to ensure the file is in a safe state when the copy is made. The destination parameter must include the name of the file. Otherwise, the command will return an error message like "Error: Wallet backup failed!" for descriptor wallets. If it is a legacy wallet, it will be copied and a file will be created with the default file name `wallet.dat`.

Reviewer 4: ```suggestion
The `backupwallet` RPC or the `Backup Wallet` GUI menu item must be used to ensure the file is in a safe state when the copy is made. The destination parameter must include the name of the file. Otherwise, the command will return an error message like "Error: Wallet backup failed!" for descriptor wallets. If it is a legacy wallet, it will be copied and a file will be created with the default file name `wallet.dat`.
```

`$ bitcoin-cli -rpcwallet="wallet-01" backupwallet /home/node01/Backups/backup-01.dat`

### 1.4 Backup Frequency

> The original Bitcoin Core wallet was a collection of unrelated private keys. If a non-HD wallet had received funds to an address and then restored from a backup made before the address was generated, then any funds sent to said address are lost forever because there's no deterministic mechanism to derive that address again.

Reviewer 4: ```suggestion
The original Bitcoin Core wallet was a collection of unrelated private keys. If a non-HD wallet had received funds to an address and then was restored from a backup made before the address was generated, then any funds sent to that address would have been lost because there was no deterministic mechanism to derive the address again.
```

> However, [version 0.13](https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.13.0.md) introduced a deterministic mechanism from which to derive keys; HD wallets. Users no longer lose funds when restoring old backups because all addresses are derived from the HD wallet seed.

Reviewer 4: ```suggestion
Bitcoin Core [version 0.13](https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.13.0.md) introduced HD wallets with deterministic key derivation. With HD wallets, users no longer lose funds when restoring old backups because all addresses are derived from the HD wallet seed.
```

> This means that a single backup is enough to recover the coins at any time. It is still recommended to make regular backups (once a week) or after a significant amount of new transactions to maintain the metadata such as labels.

Reviewer 4: ```suggestion
This means that a single backup is enough to recover the coins at any time. It is still recommended to make regular backups (once a week) or after a significant number of new transactions to maintain the metadata, such as labels.
```

> Metadata cannot be retrieved from a blockchain rescan, so if the backup is too old, the metadata will be lost forever.

Reviewer 4: It seems this sentence should be appended to the previous paragraph.

Author: Done.
